internal/server/event_handlers.go

```diff
@@ -59,6 +62,9 @@ func (s *EventServer) handleEvent() func(w http.ResponseWriter, r *http.Request)
 		alerts := make([]apiv1beta2.Alert, 0)
 	each_alert:
 		for _, alert := range allAlerts.Items {
+			alertLogger := eventLogger.WithValues("alert", client.ObjectKeyFromObject(&alert))
+			ctx := log.IntoContext(ctx, alertLogger)
```

Comment:
I'm not sure I understand the need to use the context here for carrying the logger. We can simply pass it to the called functions below. What am I missing?

Reply:
I believe this is more aligned with the context idiom. This could be beneficial if there are more functions receiving a context and using `log.FromContext()` to fetch a logger from that context to log debug info etc.

Reply:
Yeah controller runtime smells more and more like .NET/Java… Build pattern, opaque context, and soon generics. I stopped fighting it, @hiddeco seems to like it. I’m also biased, after 10 years of .NET, I moved to Go for its readability, but old patterns make their way into it, at the detriment of readability unfortunately.

Reply:
I embrace contextual loggers because we are working in a field where almost anything has a context (for the right reasons), and the tools given to us already decided to chain this together with a logger which saves us from a parameter in a code base where things often have too many of those (more than often around and beyond the 5).

I also think a logger isn't "application data", it's contextual and often not a requirement for the core logic of a function (underlined by the fact that almost any test would add some noop implementation to fulfill the requirement of the signature). Given this, I do not think that bundling a logger with a context is a wrong thing to do.

Reply:
Also, the reason for why this PR was created was because I wanted to print a debug log with useful contextual information in this [feature](https://github.com/fluxcd/notification-controller/pull/543/files#diff-9d6beb5d9cc93d6cc82ccb91135fc1f58d539aca21e2b543dfcb27c9f98e99cbR84-R103), which is probably a good example of what @hiddeco is saying in all degrees: less parameters in the `Factory` and `GooglePubSubTopic` structs and constructors, or much less parameters across all Provider implementations if changing the `Post(ctx, event)` interface to `Post(ctx, logger, event)`, or one less ugly `PostWithLogger(ctx, logger, event)` method in that interface to leave `Post(ctx, event)` untouched, and the information passed here is not as important as application data as I'd even be fine simply removing this debug log from the code (but I do see value in it and think that injecting the logger into the context wins as a non-ideal best solution). Actually, I think a logger is specifically a good example of what to inject into a context, and can't really find many others.